# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Shashank007-ux/Week-1-Run-the-Starter-Notebooks/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [4]:
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)
else:
    # find the repo root from wherever this kernel started
    while not os.path.isdir("data/raw") and os.getcwd() != "/":
        os.chdir("..")

print("Working dir:", os.getcwd())
assert os.path.exists("data/raw/content_refresh_anonymized.csv"), "starter CSV not found — are you at the repo root?"
print("Starter data found. You're ready.")

Working dir: /content/flyrank-ml-internship-starter
Starter data found. You're ready.


## 1. My lane (or freestyle) and why

I’m choosing the provisional lane: **Refresh / Content Opportunity Scoring**. I like it because the starter data already has enough observable signals to rank pages that look worth reviewing first, and the action is concrete: an editor can refresh, expand, protect, or monitor a page.

In [5]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import pandas as pd
from pathlib import Path

csv_path=Path('data/raw/content_refresh_anonymized.csv')
df=pd.read_csv(csv_path)
print(df.shape)
print(df.head(2).to_string())

(30000, 44)
             content_id          client_id  search_volume  competition competition_level   cpc     content_type    main_intent  word_count  char_count provider_used              model_used  impressions_90d  clicks_90d  pageviews_90d  sessions_90d  users_90d  engaged_sessions_90d  ai_sessions_90d  scroll_events_90d  days_with_impressions  days_with_sessions  impressions_last_30d  clicks_last_30d  sessions_last_30d  impressions_prev_30d  clicks_prev_30d  sessions_prev_30d  content_age_days age_tier  age_tier_order  days_since_last_update freshness_tier word_count_tier char_count_tier   ctr  avg_position  engagement_rate  scroll_rate  ai_traffic_pct impression_tier position_tier trend_direction  trend_pct
0  content_304f48230142  client_f369cb89fc           10.0         0.67              HIGH  2.05  keyword article  transactional      3221.0     20457.0           NaN        gemini-2.5-flash             3803          29             22            17         16                   

## 2. The question: decision, action, cost of a wrong call

My question is: **Which pages should an editor review first for refresh or protection?**

- Decision improved: which content pages deserve attention first.
- Unit of analysis: one content page.
- Output: a ranked score and a short reason code for each page.
- Action someone could take: an editor can refresh, expand, or monitor the page.
- Cost of a wrong recommendation: wasted editor time, missed opportunity, and a lower-quality review queue if weak candidates are prioritized.
- Why data or ML can help: the data contains observable signals such as demand, position, engagement, freshness, and trend, which are hard to rank by hand across thousands of pages.

In [6]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
print('Total rows:', len(df))
print('Rows with impressions_90d >= 500:', int((df['impressions_90d'] >= 500).sum()))
print('Rows with impressions_90d >= 500 and trend_direction == "down":', int(((df['impressions_90d'] >= 500) & (df['trend_direction'] == 'down')).sum()))
print('Rows with impressions_90d >= 500 and sessions_90d >= 30:', int(((df['impressions_90d'] >= 500) & (df['sessions_90d'] >= 30)).sum()))

Total rows: 30000
Rows with impressions_90d >= 500: 16726
Rows with impressions_90d >= 500 and trend_direction == "down": 9961
Rows with impressions_90d >= 500 and sessions_90d >= 30: 6869


## 3. Quick look at the data (2-3 real numbers)

1. There are **16,726** pages with at least **500** 90-day impressions.
2. There are **9,961** pages that are both high-volume and flagged as **down** by the starter trend label.
3. There are **6,869** pages that are high-volume and have at least **30** 90-day sessions, which suggests a meaningful review pool for a refresh or protection workflow.

In [7]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
summary = {
    'total_rows': len(df),
    'high_volume_pages': int((df['impressions_90d'] >= 500).sum()),
    'high_volume_and_down': int(((df['impressions_90d'] >= 500) & (df['trend_direction'] == 'down')).sum()),
    'high_volume_and_engaged': int(((df['impressions_90d'] >= 500) & (df['sessions_90d'] >= 30)).sum()),
}

summary


{'total_rows': 30000,
 'high_volume_pages': 16726,
 'high_volume_and_down': 9961,
 'high_volume_and_engaged': 6869}

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [8]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
print('This project can support a decision about which pages to review first.')
print('It can make observed, directional claims from measured signals in the starter dataset.')
print('It cannot prove that a refresh caused a recovery, and it should not claim to predict Google itself.')


This project can support a decision about which pages to review first.
It can make observed, directional claims from measured signals in the starter dataset.
It cannot prove that a refresh caused a recovery, and it should not claim to predict Google itself.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.